app/dturn/dturn.js

```diff
@@ -176,437 +181,439 @@ function sortLargestFirst(a, b) {
-                    upperStatementId,
-                    shownCount: 0,
-                    groupedCount: 0,
-                }
-                // Yes we are pushing the same object into to different lists - this is a speed optimization
-                if (!g.byLowerId[lowerStatementId]) g.byLowerId[lowerStatementId] = []
-                g.byLowerId[lowerStatementId].push(gitem)
-                if (!g.byUpperId[upperStatementId]) g.byUpperId[upperStatementId] = []
-                g.byUpperId[upperStatementId].push(gitem)
-            }
-            func(gitem)
+  if (!statementIds || statementIds.length == 0) return
+  const sortedStatementIds = statementIds.slice().sort(sortLowestIdFirst)
+  let last = sortedStatementIds.length - 1
+  if (!Discussions[discussionId].Gitems[round])
+    Discussions[discussionId].Gitems[round] = { byLowerId: {}, byUpperId: {} }
+  const g = Discussions[discussionId].Gitems[round] // shorten to speed it up
+  for (let i = 0; i <= last - 1; i++) {
+    const lowerStatementId = sortedStatementIds[i]
+    for (let j = i + 1; j <= last; j++) {
+      const upperStatementId = sortedStatementIds[j]
+      let gitem = g.byLowerId[lowerStatementId]?.find(gitem => gitem.upperStatementId == upperStatementId)
+      if (!gitem) {
+        gitem = {
+          lowerStatementId,
+          upperStatementId,
+          shownCount: 0,
+          groupedCount: 0,
         }
+        // Yes we are pushing the same object into to different lists - this is a speed optimization
+        if (!g.byLowerId[lowerStatementId]) g.byLowerId[lowerStatementId] = []
+        g.byLowerId[lowerStatementId].push(gitem)
+        if (!g.byUpperId[upperStatementId]) g.byUpperId[upperStatementId] = []
+        g.byUpperId[upperStatementId].push(gitem)
+      }
+      func(gitem)
     }
+  }
 }
 
 function putGroupings(discussionId, round, userId, groupings) {
-    const dis = Discussions[discussionId]
-    const uitem = Discussions[discussionId].Uitems[userId][round]
-    //?? if there is already a groupins, should we uncount the groupins in gitems before overriding it - in the real world groupins may get resubmitted
-    if (uitem?.groupings?.length) console.error('putGroupings already there', round, userId, groupings, uitem)
-    uitem.groupings = groupings
-    const shownStatementIds = Object.keys(uitem.shownStatementIds)
-    if (shownStatementIds.length <= 1) console.error('putGroupings', round, userId, shownStatementIds)
-    for (id of shownStatementIds) incrementShownItems(discussionId, round, id)
-    iteratePairs(discussionId, round, shownStatementIds, gitem => gitem.shownCount++)
-    groupings.forEach(group => iteratePairs(discussionId, round, group, gitem => gitem.groupedCount++))
-    Discussions[discussionId].updateUInfo({
-        [userId]: {
-            [discussionId]: { [round]: { groupings } },
-        },
-    })
+  const dis = Discussions[discussionId]
+  const uitem = Discussions[discussionId].Uitems[userId][round]
+
```

Comment:
If Discussions[disicussionId] didn't exist, line 438 would throw an error.
We don't want to throw errors at runtime.
so if(!dis) then return false
then after setting uitem the same.

Then can you make a test for calling putGroupings with a discussionId that doesn't exist.

Thanks!